CRM/Remoteevent/Registration.php

```diff
@@ -749,9 +749,32 @@ public static function registerAdditionalParticipants(RegistrationEvent $registr
         $profile = CRM_Remoteevent_RegistrationProfile::getRegistrationProfile(
            $event['event_remote_registration.remote_registration_additional_participants_profile']
         );
-        foreach ($additionalParticipants as $additionalParticipantNo => &$additionalParticipant) {
+
+        $additionalContactsData = [];
+        $additionalParticipantsData = [];
+
+        foreach ($registration->getAdditionalParticipantsData() as $additionalParticipantNo => $additionalParticipantData) {
+            $additionalParticipantsData[$additionalParticipantNo]['role_id'] = $additionalParticipantData['role_id'];
+            foreach ($profile->getFields() as $fieldKey => $fieldSpec) {
+                if (isset($additionalParticipantData[$fieldKey])) {
+                    $entity_names = (array)($fieldSpec['entity_name'] ?? $profile->getFieldEntities($fieldKey));
+                    $entity_field_name = $fieldSpec['entity_field_name'] ?? $fieldKey;
+                    $value = isset($fieldSpec['value_callback'])
+                      ? $fieldSpec['value_callback']($additionalParticipantData[$fieldKey], $additionalParticipantData)
+                      : $additionalParticipantData[$fieldKey];
+
+                    if (in_array('Contact', $entity_names, TRUE)) {
+                        $additionalContactsData[$additionalParticipantNo][$entity_field_name] = $value;
+                    }
+                    if (in_array('Participant', $entity_names, TRUE)) {
+                        $additionalParticipantsData[$additionalParticipantNo][$entity_field_name] = $value;
+                    }
+                }
+            }
+        }
```

Comment:
I think this should already be done in `RegistrationEventFactory`. This would mean that we have a separation of contact and participant data for additional participants as already for the main participant.

Reply:
I'm not sure I agree - additional participants are being handled differently with their own *CiviRemote Event* profile and also their own *XCM* profile, and should thus be processed self-contained, but maybe I'm not seeing what your proposal would actually mean …

Additional participant data is not meant to be prepopulated in the form, as those participants should not be editable at all by the "main" participant that registered them.